src/analyses/base.ml

```diff
@@ -1469,10 +1468,11 @@ struct
         if M.tracing then M.tracel "inv" "binop %a with %a %s %a == %a\n" d_exp e VD.pretty (eval e1) (show_binop op) VD.pretty (eval e2) ID.pretty c;
         (match eval e1, eval e2 with
         | `Int a, `Int b ->
-          let ik = Cilfacade.get_ikind @@ typeOf @@ if is_cmp e then e1 else e in
-          let a', b' = inv_bin_int (a, b) ik (ID.cast_to ik c) op in
-          let m1 = inv_exp (ID.cast_to (Cilfacade.get_ikind (Cil.typeOf e1)) a') e1 in
-          let m2 = inv_exp (ID.cast_to (Cilfacade.get_ikind (Cil.typeOf e2)) b') e2 in
+          let ik = Cilfacade.get_ikind @@ typeOf e1 in (* both operands have the same type! *)
```

Comment:
Why? Not true e.g. for shift in CIL

Reply:
Also, if it is a comparison, then `ik` is `CIl.IInt` regardless of the type of `e1` and `e2`

Reply:
We don't handle shift. Is it the only one?
I can revert it to `let ik = Cilfacade.get_ikind @@ typeOf @@ if is_cmp e then e1 else e in`, but this would also be wrong for shift since there's just one `ik` and it's not that of the result `e`.

Reply:
```
          match bop with
            (Mult | Div) ->
              typeMatch t1 t2; checkArithmeticType tres;
              typeMatch t1 tres; tres
          | (Eq|Ne|Lt|Le|Ge|Gt) ->
              typeMatch t1 t2; checkArithmeticType t1;
              typeMatch tres intType; tres
          | Mod|BAnd|BOr|BXor ->
              typeMatch t1 t2; checkIntegralType tres;
              typeMatch t1 tres; tres
          | LAnd | LOr ->
              checkScalarType t1; checkScalarType t2;
              typeMatch tres intType; tres
          | Shiftlt | Shiftrt ->
              typeMatch t1 tres; checkIntegralType t1;
              checkIntegralType t2; tres
          | (PlusA | MinusA) ->
                typeMatch t1 t2; typeMatch t1 tres;
                checkArithmeticType tres; tres
          | (PlusPI | MinusPI | IndexPI) ->
              checkPointerType tres;
              typeMatch t1 tres;
              checkIntegralType t2;
              tres
          | MinusPP  ->
              checkPointerType t1; checkPointerType t2;
              typeMatch t1 t2;
              typeMatch tres !ptrdiffType;
              tres

```

so `LAnd | LOr | Shiftlt | Shiftrt | (PlusPI | MinusPI | IndexPI) | MinusPP `.



Probably, it should even be the type of `e1`, since we now use `ik` like this:

```
      | Eq | Ne as op ->
        let both x = x, x in
        let m = ID.meet a b in
        (match op, ID.to_bool c with
        | Eq, Some true
        | Ne, Some false -> both m (* def. equal: if they compare equal, both values must be from the meet *)
        | Eq, Some false
        | Ne, Some true -> (* def. unequal *)
          (* Both values can not be in the meet together, but it's not sound to excude the meet from both. *)
          (* e.g. a=[0,1], b=[1,2], meet a b = [1,1], but (a != b) does not imply a=[0,0], b=[2,2] since others are possible: a=[1,1], b=[2,2] *)
          (* Only if a is a definite value, we can exclude it from b: *)
          let excl a b = match ID.to_int a with Some x -> ID.of_excl_list ik [x] | None -> b in
          meet_bin (excl b a) (excl a b)
        | _, _ -> a, b
```

Reply:
Which are probably all things we should add a comment about

Reply:
> so LAnd | LOr | Shiftlt | Shiftrt | (PlusPI | MinusPI | IndexPI).

`LAnd | LOr` can't appear since CIL transforms them into statements.
The rest we don't handle in `invariant`.

> Probably, it should even be the type of e1, since we now use ik like this:

That's what it is right now. What was your idea about `ik`? It's the type of the operands - the type of c is not of interest.

Aside: You can use ~~~ocaml instead of ~~~ for proper highlighting on github.

Reply:
> What was your idea about ik? It's the type of the operands - the type of c is not of interest.

I found it necessary to cast `c` to `ik` as well at some point for handling more complicated expressions that involve several levels of arithmetic operations. But I'm not sure if it still is required.

> Aside: You can use ~~~ocaml instead of ~~~ for proper highlighting on github.

I didn't know it works for OCaml to (and not only C), thx!